test: p2p_feefilter improvements (logging, refactoring, speedup)

theStack opened the pull request:
This PR gives some love to the functional test `p2p_feefilter.py` by introducing the following changes:
* add missing log messages for the `test_feefilter` subtest (the main one)
* deduplicate code by using the utility function `wait_until` (already using the [recently introduced version](https://github.com/bitcoin/bitcoin/commit/12410b1feb80189061eb4a2b43421e53cbb758a8)) instead of a manual condition checking loop with `time.sleep`
* improve naming of the function `matchAllInvs` (more expressive name, snake_case) and moving it from global namespace to the connection class `FeefilterConn`
* speeding up the test significantly by the good ol' method of activating immediate tx relay (as seen on e.g. https://github.com/bitcoin/bitcoin/pull/17121, https://github.com/bitcoin/bitcoin/pull/17124, https://github.com/bitcoin/bitcoin/pull/17340, https://github.com/bitcoin/bitcoin/pull/17362, ...):
```
master branch:
$ time ./p2p_feefilter.py
...
real    0m39.367s
user    0m1.227s
sys 0m0.571s

PR branch:
$ time ./p2p_feefilter.py
...
real    0m9.386s
user    0m1.120s
sys 0m0.577s
```

theStack committed:
test: add logging for p2p_feefilter.py

theStack committed:
test: use wait_until for invs matching in p2p_feefilter.py

additionally:
    -> rename function with snake_case (s/allInvsMatch/wait_for_invs_to_match)
    -> move it from global namespace to the class FeefilterConn

theStack committed:
test: speedup p2p_feefilter.py by whitelisting peers (immediate tx relay)

Most of the test time is spent in wait_for_invs() after sending to addresses,
i.e. the bottleneck is in relaying transactions. By whitelisting the peers via
-whitelist, the inventory is transmissioned immediately rather than on average
every 5 seconds, speeding up the test significantly:

before:
$ time ./p2p_feefilter.py
...
real    0m39.367s
user    0m1.227s
sys 0m0.571s

with this commit:
$ time ./p2p_feefilter.py
...
real    0m9.386s
user    0m1.120s
sys 0m0.577s

jonatack commented on test/functional/p2p_feefilter.py:
```diff
@@ -88,24 +88,22 @@ def test_feefilter(self):
         conn.clear_invs()
 
-        # Set a filter of .15 sat/byte on test connection
+        self.log.info("Set a fee filter of .15 sat/byte on test connection")
```
I think you can leave this line as a code comment instead of logging it.

For the logging in general, would you mind prefixing the decimal values with 0? e.g. s/.15/0.15/

This is personal taste, so feel free to ignore, but I think it's unnecessarily verbose to add "that" after "Test..."

theStack replied:
Done.

jonatack commented on test/functional/p2p_feefilter.py:
```diff
@@ -48,6 +37,12 @@ def on_inv(self, message):
 
+    def wait_for_invs(self, test_framework, invs_expected):
+        def invs_match():
+            return sorted(invs_expected) == sorted(self.txinvs)
```
could inline this with a lambda?

theStack replied:
Before I had done it using a lambda and an assignment, the the following way:
```
invs_match = lambda: sorted(invs_expected) == sorted(self.txinvs)
test_framework.wait_until(invs_match, timeout=60, lock=mininode_lock)
```
but then flake8 would give the following error:
`p2p_feefilter.py:41:9: E731 do not assign a lambda expression, use a def`

Directly using the lambda on place of the parameter of `wait_until` works though, done.

jonatack commented on test/functional/p2p_feefilter.py:
```diff
@@ -48,6 +37,12 @@ def on_inv(self, message):
             if (i.type == MSG_TX):
                 self.txinvs.append(hashToHex(i.hash))
 
+    def wait_for_invs(self, test_framework, invs_expected):
```
perhaps call this `assert_invs_match` or `wait_for_invs_to_match`

theStack replied:
Done, I chose `wait_for_invs_to_match`.

jonatack commented on test/functional/p2p_feefilter.py:
```diff
@@ -57,6 +57,9 @@ def set_test_params(self):
         self.extra_args = [["-minrelaytxfee=0.00000100", "-mintxfee=0.00000100"]] * self.num_nodes
+        # whitelist all peers to speed up tx relay / mempool sync
+        for args in self.extra_args:
+            args.append("-whitelist=noban@127.0.0.1")
```
could just inline this arg with the others and put the comment above line 59

theStack replied:
Done.

jonatack reviewed (commented):
Concept ACK

Edit: verified the test runs significantly more quickly with the noban arg.

theStack commented:
Thans a lot for reviewing jonatack! Force-pushed with your suggested changes.

jonatack commented on test/functional/p2p_feefilter.py:
```diff
@@ -48,6 +37,10 @@ def on_inv(self, message):
                 self.txinvs.append(hashToHex(i.hash))
 
+    def wait_for_invs_to_match(self, test_framework, invs_expected):
+        test_framework.wait_until(lambda: sorted(invs_expected) == sorted(self.txinvs),
```
I wonder if we can eek some more perf by memoizing (untested):
```suggestion
        expected_txinvs = sorted(invs_expected)
        test_framework.wait_until(lambda: sorted(self.txinvs) == expected_txinvs),
```

theStack replied:
Yes, that makes to do the sorting of the txinvs to test for only once. Did it with in-place sort of the passed list.

jonatack commented on test/functional/p2p_feefilter.py:
```diff
@@ -61,7 +54,9 @@ def set_test_params(self):
         # rounding down 3 places, leading to stranded transactions.
         # See issue #16499
-        self.extra_args = [["-minrelaytxfee=0.00000100", "-mintxfee=0.00000100"]] * self.num_nodes
+        # whitelist all peers to speed up tx relay / mempool sync
```
Sorry I didn't notice this earlier; I think we're now starting to call this s/whitelist all/grant noban permission to/ (or something similar)

theStack replied:
Thanks, done.

jonatack reviewed (commented):
LGTM, a couple more comments below.

theStack commented:
Rebased on master and added suggested changes by jonatack (only sort expected invs once, use correct terminology regarding noban permission terminology).

jonatack commented on test/functional/p2p_feefilter.py:
```diff
@@ -88,24 +84,22 @@ def test_feefilter(self):
         conn.send_and_ping(msg_feefilter(150))
 
-        # Test that txs are still being received by test connection (paying .15 sat/byte)
+        self.log.info("Test txs paying 0.15 sat/byte are being received by test connection")
```
c063dce9 nit: if you have to retouch, can remove "being" here and in line 87.

jonatack reviewed (commented):
ACK 2d43d84

Restarted the stalled s390x Travis job.

theStack commented:
Rebased.

jonatack commented:
re-ACK ea74a3c per `git range-diff ce3bdd0e 2d43d84 ea74a3c`

MarcoFalke commented on test/functional/p2p_feefilter.py:
```diff
@@ -48,6 +37,11 @@ def on_inv(self, message):
 
+    def wait_for_invs_to_match(self, test_framework, invs_expected):
+        invs_expected.sort()
+        test_framework.wait_until(lambda: invs_expected == sorted(self.txinvs),
```
why not self.wait_until?

theStack replied:
Good idea, this makes the code a lot shorter -- `test_framework` doesn't have to be passed to the inv-wait function and also setting the lock parameter is not needed anymore.
Done.

MarcoFalke commented on test/functional/p2p_feefilter.py:
```diff
@@ -61,7 +55,9 @@ def set_test_params(self):
-        self.extra_args = [["-minrelaytxfee=0.00000100", "-mintxfee=0.00000100"]] * self.num_nodes
+        # grant noban permission to all peers to speed up tx relay / mempool sync
+        self.extra_args = [["-minrelaytxfee=0.00000100", "-mintxfee=0.00000100",
+                            "-whitelist=noban@127.0.0.1"]] * self.num_nodes
```
would be nice to either put all in one line or each in one new line

theStack replied:
Decided for the one-arg-per-line version. Done.

maflcko reviewed (approved):
Looks fine, some style nits


4fbda7da50d508481d9a94b00e4dc994dad9465d

theStack commented:
Thanks for reviewing Marco. Force-pushed with your suggestions and a small log-message change by jonatack  previously (https://github.com/bitcoin/bitcoin/pull/19564#discussion_r465517818).

instagibbs commented:
code review ACK https://github.com/bitcoin/bitcoin/pull/19564/commits/9e7894357e296dbe0be775e1527654729dbb71b0

jonatack commented:
re-ACK 9e78943 per `git range-diff 3ab2582 ea74a3c 9e78943`
